File: gui/src/catalog.ts
// A faithful TypeScript port of `allmystuff-graph`'s routing +
// authorization rules. The desktop app runs this client-side so the graph
// is fully interactive on its own; when the Tauri backend is present it
// validates the same way in Rust before anything touches the wire. Two
// implementations, one set of rules — kept honest by sharing the exact
// shapes in `types.ts`.

import type {
  Capability,
  Catalog,
  Flow,
  Grant,
  GrantRole,
  MediaKind,
  Route,
} from "./types";
import { MEDIA } from "./types";

export type ConnectResult =
  | { ok: true; route: Route }
  | { ok: false; reason: string; denied?: GrantRequest[] };

export interface GrantRequest {
  node: string;
  person: string;
  personName: string;
  media: MediaKind;
  role: GrantRole;
  capability: string | null;
  description: string;
}

export function capability(cat: Catalog, id: string): Capability | undefined {
  return cat.capabilities.find((c) => c.id === id);
}

/** Resolve a capability *for display*, synthesizing a stand-in for the
 *  terminal and files endpoints that are deliberately never in the catalog
 *  (a persistent generic capability would match every auto-wiring picker —
 *  `matchEndpoint` treats generic as compatible with everything). This
 *  keeps a live terminal/files session visible in "Connected now" and on
 *  the graph without making it a wireable thing. */
export function capabilityForDisplay(cat: Catalog, id: string): Capability | undefined {
  const real = capability(cat, id);
  if (real) return real;
  const sep = id.indexOf(":");
  if (sep < 0) return undefined;
  const node = id.slice(0, sep);
  const tail = id.slice(sep + 1);
  if (tail === "terminal")
    return { id, node, label: "Terminal", media: "generic", flow: "source", origin: "terminal" };
  if (tail.startsWith("term-view"))
    return {
      id,
      node,
      label: "Terminal viewer",
      media: "generic",
      flow: "sink",
      origin: "terminal-view",
    };
  if (tail === "files")
    return { id, node, label: "Files", media: "generic", flow: "source", origin: "files" };
  if (tail.startsWith("files-view"))
    return {
      id,
      node,
      label: "Files viewer",
      media: "generic",
      flow: "sink",
      origin: "files-view",
    };
  return undefined;
}

export function mediaCompatible(a: MediaKind, b: MediaKind): boolean {
  return a === b || a === "generic" || b === "generic";
}

export function canSource(f: Flow): boolean {
  return f === "source" || f === "duplex";
}
export function canSink(f: Flow): boolean {
  return f === "sink" || f === "duplex";
}

function grantPermits(g: Grant, media: MediaKind, role: GrantRole, capId: string): boolean {
  if (!mediaCompatible(g.media, media)) return false;
  if (g.capability && g.capability !== capId) return false;
  if (role === "provide") return g.role === "provide" || g.role === "both";
  if (role === "consume") return g.role === "consume" || g.role === "both";
  return g.role === "both";
}

/** The content-derived, stable id for a grant of this scope in `person`'s
 *  share — `grant:{person}:{media}:{role}:{capability|*}`. Mirrors
 *  `Grant::id_for` in `allmystuff-graph` (model.rs) byte-for-byte: two
 *  structurally identical grants collapse to one id, and the id is identical
 *  across a restart and on both peers, so persistence, de-dupe, and
 *  revoke-by-id all agree. Change both together.
 *
 *  `media` and `role` are already the snake_case wire tokens here
 *  (`generic`, `provide`, …), matching `MediaKind::token` / `GrantRole::label`
 *  on the Rust side. */
export function scopedGrantId(
  person: string,
  media: MediaKind,
  role: GrantRole,
  capability: string | null,
): string {
  return `grant:${person}:${media}:${role}:${capability ?? "*"}`;
}

/** Returns a GrantRequest if the endpoint is on a shared node lacking
 *  coverage, otherwise null (mine, or already granted). A grant authorizes
 *  the *person*, not one machine — sharing with someone lets them route
 *  the granted thing to any of their nodes — so coverage is the union of
 *  grants across every node shared with the same person (mirrors the Rust
 *  `Catalog::check_endpoint`). */
function checkEndpoint(
  cat: Catalog,
  capId: string,
  media: MediaKind,
  role: GrantRole,
): GrantRequest | null {
  const cap = capability(cat, capId);
  if (!cap) return null;
  const node = cat.nodes.find((n) => n.id === cap.node);
  if (!node || node.relationship.kind !== "shared") return null;
  const share = node.relationship;
  const ok = cat.nodes.some(
    (n) =>
      n.relationship.kind === "shared" &&
      n.relationship.person.id === share.person.id &&
      n.relationship.grants.some((g) => grantPermits(g, media, role, capId)),
  );
  if (ok) return null;
  return {
    node: cap.node,
    person: share.person.id,
    personName: share.person.name,
    media,
    role,
    capability: capId,
    description: describeGrant(media, role),
  };
}

export function describeGrant(media: MediaKind, role: GrantRole): string {
  const m = MEDIA[media].label.toLowerCase();
  if (role === "provide") return `Send their ${m}`;
  if (role === "consume") return `Receive your ${m}`;
  return `Exchange ${m}`;
}

export function describeAction(media: MediaKind, role: GrantRole): string {
  const m = MEDIA[media].label.toLowerCase();
  if (role === "provide") return `send you their ${m}`;
  if (role === "consume") return `receive your ${m}`;
  return `exchange ${m} with you`;
}

export function requiredGrants(cat: Catalog, from: string, to: string): GrantRequest[] {
  const media = capability(cat, from)?.media ?? "generic";
  const out: GrantRequest[] = [];
  const a = checkEndpoint(cat, from, media, "provide");
  const b = checkEndpoint(cat, to, media, "consume");
  if (a) out.push(a);
  if (b) out.push(b);
  return out;
}

export function routeId(from: string, to: string): string {
  return `route:${from}→${to}`;
}

/** Validate + authorize a connection. Mirrors `Catalog::propose_route`. */
export function proposeRoute(cat: Catalog, from: string, to: string): ConnectResult {
  const structural = validateRoute(cat, from, to);
  if (!structural.ok) return structural;

  const denied = requiredGrants(cat, from, to);
  if (denied.length > 0) {
    const who = denied[0].personName;
    const act = describeAction(denied[0].media, denied[0].role);
    return { ok: false, reason: `${who} isn't allowed to ${act} yet.`, denied };
  }
  return structural;
}

/** Validate a connection for the **rooms plane** — every structural rule
 *  of `proposeRoute` without the share-grant gate. Being in the same room
 *  *is* the consent, and it's scoped to the room session: the route lives
 *  only while the room toggle does, and no standing grant is ever minted
 *  for it. Mirrors `Catalog::propose_room_route`. */
export function proposeRoomRoute(cat: Catalog, from: string, to: string): ConnectResult {
  return validateRoute(cat, from, to);
}

/** The structural half of a proposal (no authorization). */
function validateRoute(cat: Catalog, from: string, to: string): ConnectResult {
  if (from === to) return { ok: false, reason: "A thing can't connect to itself." };
  const src = capability(cat, from);
  const dst = capability(cat, to);
  if (!src) return { ok: false, reason: `Unknown capability: ${from}` };
  if (!dst) return { ok: false, reason: `Unknown capability: ${to}` };
  // An unclaimed device is on the mesh but not yet yours — claim it (or mark
  // it shared) before anything can route to or from it.
  const srcNode = cat.nodes.find((n) => n.id === src.node);
  const dstNode = cat.nodes.find((n) => n.id === dst.node);
  if (srcNode?.relationship.kind === "unclaimed")
    return { ok: false, reason: `${srcNode.label} isn't yours yet — claim it first.` };
  if (dstNode?.relationship.kind === "unclaimed")
    return { ok: false, reason: `${dstNode.label} isn't yours yet — claim it first.` };
  if (!canSource(src.flow))
    return { ok: false, reason: `${src.label} can't send — it only receives.` };
  if (!canSink(dst.flow))
    return { ok: false, reason: `${dst.label} can't receive — it only sends.` };
  if (!mediaCompatible(src.media, dst.media))
    return {
      ok: false,
      reason: `${MEDIA[src.media].label} doesn't fit ${MEDIA[dst.media].label}.`,
    };

  const media = src.media !== "generic" ? src.media : dst.media;
  return { ok: true, route: { id: routeId(from, to), from, to, media } };
}

const MACHINE_ORIGINS = new Set(["screen", "control", "controller", "system", "viewer"]);

/** Sort key (lower = preferred), mirroring the Rust `endpoint_rank`: a
 *  synthetic machine endpoint first, then the category's current default
 *  device, then everything else. */
function endpointRank(c: Capability): number {
  if (MACHINE_ORIGINS.has(c.origin)) return 0;
  if (c.default) return 1;
  return 2;
}

export function matchEndpoint(
  cat: Catalog,
  node: string,
  media: MediaKind,
  role: GrantRole,
): Capability | undefined {
  return cat.capabilities
    .filter((c) => c.node === node && mediaCompatible(c.media, media))
    .filter((c) => (role === "provide" ? canSource(c.flow) : canSink(c.flow)))
    .sort((a, b) => endpointRank(a) - endpointRank(b) || a.id.localeCompare(b.id))[0];
}

File: crates/allmystuff-graph/src/catalog.rs
//! The [`Catalog`] — the whole graph in one place: which nodes exist,
//! what each can do, what's wired to what, and which bundles exist. It's
//! the single object the UI reads and the only thing that mints a
//! [`Route`], because route creation is exactly where media, flow, and
//! authorization get enforced.

use serde::{Deserialize, Serialize};

use crate::authz::{describe_action, describe_grant, Denied, GrantRequest};
use crate::model::*;

/// Everything the graph knows. Cheap to clone for a snapshot; the GUI
/// holds one and re-renders from it.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Catalog {
    pub nodes: Vec<MeshNode>,
    pub capabilities: Vec<Capability>,
    pub routes: Vec<Route>,
    pub groups: Vec<Group>,
}

/// Why a connection couldn't be made.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum ConnectError {
    #[error("unknown capability: {0}")]
    UnknownCapability(String),
    #[error("unknown node: {0}")]
    UnknownNode(String),
    #[error("unknown group: {0}")]
    UnknownGroup(String),
    #[error("a capability can't connect to itself")]
    SelfLoop,
    #[error("{from} produces {from_media} but {to} expects {to_media}")]
    MediaMismatch {
        from: String,
        to: String,
        from_media: &'static str,
        to_media: &'static str,
    },
    #[error("{cap} can't act as a {wanted} ({label} is {flow})")]
    WrongFlow {
        cap: String,
        label: String,
        wanted: &'static str,
        flow: &'static str,
    },
    #[error("nothing on {node} can {verb} {media} to complete this group")]
    NoMatchingEndpoint {
        node: String,
        media: &'static str,
        verb: &'static str,
    },
    /// Boxed because [`Denied`] is by far the largest variant; boxing it
    /// keeps `Result<Route, ConnectError>` small on the happy path.
    #[error(transparent)]
    Denied(Box<Denied>),
}

impl From<Denied> for ConnectError {
    fn from(d: Denied) -> Self {
        ConnectError::Denied(Box::new(d))
    }
}

impl From<Box<Denied>> for ConnectError {
    fn from(d: Box<Denied>) -> Self {
        ConnectError::Denied(d)
    }
}

impl Catalog {
    pub fn new() -> Self {
        Self::default()
    }

    // ---- lookups ------------------------------------------------------

    pub fn node(&self, id: &NodeId) -> Option<&MeshNode> {
        self.nodes.iter().find(|n| &n.id == id)
    }

    pub fn capability(&self, id: &CapabilityId) -> Option<&Capability> {
        self.capabilities.iter().find(|c| &c.id == id)
    }

    pub fn group(&self, id: &str) -> Option<&Group> {
        self.groups.iter().find(|g| g.id == id)
    }

    /// Capabilities physically present on a node.
    pub fn capabilities_on<'a>(
        &'a self,
        node: &'a NodeId,
    ) -> impl Iterator<Item = &'a Capability> + 'a {
        self.capabilities.iter().filter(move |c| &c.node == node)
    }

    fn require_cap(&self, id: &CapabilityId) -> Result<&Capability, ConnectError> {
        self.capability(id)
            .ok_or_else(|| ConnectError::UnknownCapability(id.to_string()))
    }

    fn require_node(&self, id: &NodeId) -> Result<&MeshNode, ConnectError> {
        self.node(id)
            .ok_or_else(|| ConnectError::UnknownNode(id.to_string()))
    }

    // ---- single-route creation ---------------------------------------

    /// Validate and authorize a connection from one capability to
    /// another, returning the [`Route`] it would create (not yet added to
    /// the catalog — call [`Catalog::add_route`] to commit it). This is
    /// the one place media, flow, and authorization are all checked.
    pub fn propose_route(
        &self,
        from: &CapabilityId,
        to: &CapabilityId,
    ) -> Result<Route, ConnectError> {
        if from == to {
            return Err(ConnectError::SelfLoop);
        }
        let src = self.require_cap(from)?;
        let dst = self.require_cap(to)?;

        if !src.flow.can_source() {
            return Err(ConnectError::WrongFlow {
                cap: src.id.to_string(),
                label: src.label.clone(),
                wanted: "source",
                flow: flow_word(src.flow),
            });
        }
        if !dst.flow.can_sink() {
            return Err(ConnectError::WrongFlow {
                cap: dst.id.to_string(),
                label: dst.label.clone(),
                wanted: "sink",
                flow: flow_word(dst.flow),
            });
        }
        if !src.media.compatible(dst.media) {
            return Err(ConnectError::MediaMismatch {
                from: src.id.to_string(),
                to: dst.id.to_string(),
                from_media: src.media.label(),
                to_media: dst.media.label(),
            });
        }

        // Media for the route: the concrete one if either side pins it,
        // else Generic.
        let media = if src.media != MediaKind::Generic {
            src.media
        } else {
            dst.media
        };

        let route = Route {
            id: route_id(from, to),
            from: from.clone(),
            to: to.clone(),
            media,
            group: None,
        };
        self.authorize(&route)?;
        Ok(route)
    }

    /// Check a route against the relationships in play. Endpoints on your
    /// own nodes always pass; endpoints on a shared node need a grant for
    /// the role they're playing (the `from` side provides, the `to` side
    /// consumes).
    pub fn authorize(&self, route: &Route) -> Result<(), Box<Denied>> {
        self.check_endpoint(&route.from, route.media, GrantRole::Provide)?;
        self.check_endpoint(&route.to, route.media, GrantRole::Consume)?;
        Ok(())
    }

    fn check_endpoint(
        &self,
        cap_id: &CapabilityId,
        media: MediaKind,
        role: GrantRole,
    ) -> Result<(), Box<Denied>> {
        // A capability we can't find, or one on a node we don't model, is
        // treated as ours — there's no *shared* relationship to gate it.
        let Some(cap) = self.capability(cap_id) else {
            return Ok(());
        };
        let Some(node) = self.node(&cap.node) else {
            return Ok(());
        };
        let Some(share) = node.relationship.share() else {
            return Ok(()); // Mine — always allowed.
        };
        let permitted = share.grants.iter().any(|g| g.permits(media, role, cap_id));
        if permitted {
            Ok(())
        } else {
            Err(Box::new(Denied {
                node: cap.node.clone(),
                person: share.person.id.clone(),
                person_name: share.person.name.clone(),
                media,
                role,
                capability: cap_id.clone(),
                action: describe_action(media, role),
            }))
        }
    }

    /// For a denied (or hypothetical) route, the minimal grant(s) that
    /// would authorize it — one per shared endpoint that's missing
    /// coverage. Powers the "Let {person} …?" one-tap share prompt.
    pub fn required_grants(&self, from: &CapabilityId, to: &CapabilityId) -> Vec<GrantRequest> {
        let media = self
            .capability(from)
            .map(|c| c.media)
            .unwrap_or(MediaKind::Generic);
        let mut out = Vec::new();
        for (cap_id, role) in [(from, GrantRole::Provide), (to, GrantRole::Consume)] {
            if let Err(denied) = self.check_endpoint(cap_id, media, role) {
                let denied = *denied;
                out.push(GrantRequest {
                    node: denied.node,
                    person: denied.person,
                    person_name: denied.person_name,
                    media: denied.media,
                    role: denied.role,
                    capability: Some(denied.capability),
                    description: describe_grant(media, role),
                });
            }
        }
        out
    }

    /// Commit a route to the catalog (idempotent on route id).
    pub fn add_route(&mut self, route: Route) {
        if !self.routes.iter().any(|r| r.id == route.id) {
            self.routes.push(route);
        }
    }

    /// Remove a route by id. Returns whether anything was removed.
    pub fn remove_route(&mut self, id: &str) -> bool {
        let before = self.routes.len();
        self.routes.retain(|r| r.id != id);
        self.routes.len() != before
    }

    // ---- group fan-out ------------------------------------------------

    /// Wire every member of a group to `target` as a unit. Each member
    /// becomes one route in the direction its flow implies:
    ///
    ///  * a **source** member (mic, keyboard, this screen) feeds the
    ///    target's matching sink;
    ///  * a **sink** member (monitor, speaker) is fed by the target's
    ///    matching source;
    ///  * a **duplex** member wires both directions that exist.
    ///
    /// Members with no counterpart on the target are skipped (you can't
    /// route a mic to a machine that records nothing) — but an
    /// authorization failure is never skipped; it aborts the whole
    /// connect so a group can't partially breach a share.
    pub fn connect_group(
        &self,
        group_id: &str,
        target: &NodeId,
    ) -> Result<Vec<Route>, ConnectError> {
        let group = self
            .group(group_id)
            .ok_or_else(|| ConnectError::UnknownGroup(group_id.to_string()))?;
        self.require_node(target)?;
        if &group.node == target {
            return Err(ConnectError::SelfLoop);
        }

        let mut routes = Vec::new();
        for member_id in &group.members {
            let Some(member) = self.capability(member_id) else {
                continue;
            };
            // Outbound leg: member sources → a target sink.
            if member.flow.can_source() {
                if let Some(sink) = self.match_endpoint(target, member.media, GrantRole::Consume) {
                    routes.push(self.group_route(member_id, &sink.id, member.media, group_id)?);
                }
            }
            // Inbound leg: a target source → member sinks.
            if member.flow.can_sink() {
                if let Some(src) = self.match_endpoint(target, member.media, GrantRole::Provide) {
                    routes.push(self.group_route(&src.id, member_id, member.media, group_id)?);
                }
            }
        }

        if routes.is_empty() {
            // Report the first member's unmet need so the UI can explain.
            let first = group
                .members
                .first()
                .and_then(|m| self.capability(m))
                .map(|c| (c.node.to_string(), c.media.label()))
                .unwrap_or_default();
            return Err(ConnectError::NoMatchingEndpoint {
                node: target.to_string(),
                media: leak_media_label(first.1),
                verb: "exchange",
            });
        }
        Ok(routes)
    }

    /// Pick the target capability that can play `role` for `media`. Prefers
    /// a synthetic machine capability (origin `screen`/`control`/`system`)
    /// so an RDC group lands on "this computer" rather than a stray device,
    /// then falls back to the first physical match by id order.
    fn match_endpoint(
        &self,
        node: &NodeId,
        media: MediaKind,
        role: GrantRole,
    ) -> Option<&Capability> {
        // Clone the node id so the returned reference is tied to `&self`
        // alone, not to the (shorter-lived) `node` argument.
        let node = node.clone();
        let mut candidates: Vec<&Capability> = self
            .capabilities
            .iter()
            .filter(|c| c.node == node)
            .filter(|c| c.media.compatible(media))
            .filter(|c| match role {
                GrantRole::Provide => c.flow.can_source(),
                GrantRole::Consume => c.flow.can_sink(),
                GrantRole::Both => c.flow == Flow::Duplex,
            })
            .collect();
        candidates.sort_by(|a, b| {
            let rank = |c: &Capability| u8::from(!is_machine_origin(&c.origin));
            rank(a).cmp(&rank(b)).then_with(|| a.id.0.cmp(&b.id.0))
        });
        candidates.into_iter().next()
    }

    fn group_route(
        &self,
        from: &CapabilityId,
        to: &CapabilityId,
        media: MediaKind,
        group_id: &str,
    ) -> Result<Route, ConnectError> {
        let route = Route {
            id: route_id(from, to),
            from: from.clone(),
            to: to.clone(),
            media,
            group: Some(group_id.to_string()),
        };
        // Authorization still applies to every leg.
        self.authorize(&route)?;
        Ok(route)
    }
}

fn route_id(from: &CapabilityId, to: &CapabilityId) -> String {
    format!("route:{from}→{to}")
}

fn flow_word(f: Flow) -> &'static str {
    match f {
        Flow::Source => "a source",
        Flow::Sink => "a sink",
        Flow::Duplex => "duplex",
    }
}

fn is_machine_origin(origin: &str) -> bool {
    matches!(origin, "screen" | "control" | "system")
}

// `MediaKind::label` returns `&'static str`; this just re-borrows the
// matching static so `NoMatchingEndpoint` can hold it.
fn leak_media_label(s: &str) -> &'static str {
    for m in [
        MediaKind::Audio,
        MediaKind::Video,
        MediaKind::Display,
        MediaKind::Input,
        MediaKind::Storage,
        MediaKind::Generic,
    ] {
        if m.label() == s {
            return m.label();
        }
    }
    "data"
}
